### PR TITLE
[TBD] Systray

### PR DESCRIPTION
**Pull request by jmt85**

### CODE REVIEW

**Comment by julianseeger** [comment]
Tabs please

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 1]
```diff
@@ -21,110 +21,150 @@
...
+    private void setTrayIcon(Stage primaryStage) throws ClassNotFoundException, UnsupportedLookAndFeelException, InstantiationException, IllegalAccessException {
+
+
+        SystemTray sTray = SystemTray.getSystemTray();
+        primaryStage.setOnCloseRequest(arg0 -> primaryStage.hide());
+
+        final JPopupMenu popup = buildSystemTrayJPopupMenu(primaryStage);
+
+
+
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: please take care of unneccessary blank lines, some of them increase readability but too many decrease it

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 2]
```diff
@@ -21,110 +21,150 @@
...
+            boolean visible = false;
+
+            @Override
+            public void mouseReleased(MouseEvent e) {
+                if (e.getButton() == MouseEvent.BUTTON1) {
+                    if (!visible) {
+                        popup.setLocation(e.getX(), e.getY() + 10);
+                        popup.setVisible(true);
+                        visible = true;
+                        visible = true;
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: sure? ;)

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 3]
```diff
@@ -21,110 +21,150 @@
...
+
+    private void trayIconListener(final JPopupMenu popup, TrayIcon icon) {
+        icon.addMouseListener(new MouseAdapter() {
+            boolean visible = false;
+
+            @Override
+            public void mouseReleased(MouseEvent e) {
+                if (e.getButton() == MouseEvent.BUTTON1) {
+                    if (!visible) {
+                        popup.setLocation(e.getX(), e.getY() + 10);
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: This is useless if the taskbar is on the bottom of the screen or the right.
Windows default is bottom so this won't work

File: src/main/java/de/qabel/desktop/DesktopClient.java (hunk 2 above)
Comment (jmt85) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: savety first

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 4]
```diff
@@ -21,110 +21,150 @@
...
+                if (arg instanceof Account) {
+                    Scene layoutScene = new Scene(new LayoutView().getView(), 800, 600);
+                    Platform.runLater(() -> primaryStage.setScene(layoutScene));
+                }
+            });
+        } else {
+            scene = new Scene(new LayoutView().getView(), 800, 600);
+        }
+
+        Platform.setImplicitExit(false);
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: I think you need a shutdown handler with `Platform.exit()` to shutdown properly.

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 5]
```diff
@@ -21,110 +21,150 @@
...
+        SystemTray sTray = SystemTray.getSystemTray();
+        primaryStage.setOnCloseRequest(arg0 -> primaryStage.hide());
+
+        final JPopupMenu popup = buildSystemTrayJPopupMenu(primaryStage);
+
+
+
+        URL url = System.class.getResource("/logo.png");
+        Image img = Toolkit.getDefaultToolkit().getImage(url);
+        TrayIcon icon = new TrayIcon(img, "Qabel", null);
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: you can omit the 3rd argument (there is a constructor `TrayIcon(Image image, String tooltip)` )

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 6]
```diff
@@ -21,110 +21,150 @@
...
+                System.exit(0);
+            }
+        };
+
+        for (JMenuItem item : new JMenuItem[]{showMenuItem, exitMenuItem}) {
+            if (item == exitMenuItem) menu.addSeparator();
+            menu.add(item);
+            item.addActionListener(listener);
+        }
+        return menu;
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: pretty complicated.
```JAVA
menu.add(showMenuItem);
menu.add(exitMenuItem);
showMenuItem.addActionListener(ae -> Platform.runLater(primaryStage::show));
exitMenuItem.addActionListener(ae -> System.exit(0));
```
should do the same...

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 7]
```diff
@@ -32,10 +34,10 @@
 
 public class RemoteFSController extends AbstractController implements Initializable {
 
-    private static final String PRIVATE_KEY = "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a";
-    public static final String ROOT_FOLDER_NAME = "root Folder";
-    private final String bucket = "qabel";
-    private final String prefix = "qabelTest";
+    private String PRIVATE_KEY;
+    public String ROOT_FOLDER_NAME = "RootFolder";
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: If they are not static, then [LCC](https://en.wikipedia.org/wiki/CamelCase#Variations_and_synonyms) please.
Additionally, no public properties please. Package private if you need it for tests, protected if they should be part of the inheritance-API and private by default.

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 8]
```diff
@@ -48,10 +50,22 @@
...
     public void initialize(URL location, ResourceBundle resources) {
 
+        if (clientConfiguration.hasAccount()) {
+            PRIVATE_KEY = clientConfiguration.getSelectedIdentity().getPrimaryKeyPair().getPrivateKey().toString();
+            PRIVATE_KEY = "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a";
+        } else {
+            PRIVATE_KEY = "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a";
+        }
+
+        System.out.print(PRIVATE_KEY);
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: ?
* hasAccount() is not related to getSelectedIdentity()
* the `if` has no effect ... result is always `PRIVATE_KEY = "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a";`
* Debug output...

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 9]
```diff
@@ -128,11 +142,26 @@ protected void handleUploadFolderButtonAction(ActionEvent event) throws QblStora
     }
 
     @FXML
+    protected void handleDownloadButtonAction(ActionEvent event) throws QblStorageException, IOException {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: Mixed with #89 ?

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 10]
```diff
@@ -123,6 +114,44 @@ private void setTrayIcon(Stage primaryStage) {
 
 	}
 
+	private void trayIconListener(final JPopupMenu popup, TrayIcon icon) {
+		icon.addMouseListener(new MouseAdapter() {
+			boolean visible = false;
+
+			@Override
+			public void mouseReleased(MouseEvent e) {
+				if (e.getButton() == MouseEvent.BUTTON1) {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: Just a hint:
the GuardClause-Pattern fits perfectly to reduce nesting:
```JAVA
if (e.getButton() != MouseEvent.BUTTON1) {
   return; // now everyone knows you are only interested in BUTTON1
}
... action handling ...
```

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 11]
```diff
@@ -123,6 +114,44 @@ private void setTrayIcon(Stage primaryStage) {
...
+					if (!visible) {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: shortened version:
```JAVA
popup.setLocation(e.getX(), e.getY());
visible = !visible;
popup.setVisible(visible);
```
(no ifs needed)

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 12]
```diff
@@ -123,6 +114,44 @@ private void setTrayIcon(Stage primaryStage) {
...
+						popup.setLocation(e.getX(), e.getY());
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: still not windows-safe ;)

File: src/main/java/de/qabel/desktop/DesktopClient.java [hunk 13]
```diff
@@ -123,6 +114,44 @@ private void setTrayIcon(Stage primaryStage) {
...
+		final JMenuItem showMenuItem = new JMenuItem("Show");
+		final JMenuItem exitMenuItem = new JMenuItem("Exit");
+
+		menu.add(showMenuItem);
+		menu.add(exitMenuItem);
+		showMenuItem.addActionListener(ae -> Platform.runLater(primaryStage::show));
+		exitMenuItem.addActionListener(ae -> System.exit(0));
+
+
+		for (JMenuItem item : new JMenuItem[]{showMenuItem, exitMenuItem}) {
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/DesktopClient.java]: you are adding the menu items twice now

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 14]
```diff
@@ -32,10 +41,10 @@
 
 public class RemoteFSController extends AbstractController implements Initializable {
 
-    private static final String PRIVATE_KEY = "77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a";
-    public static final String ROOT_FOLDER_NAME = "root Folder";
-    private final String bucket = "qabel";
-    private final String prefix = "qabelTest";
+    private QblECKeyPair KEY_PAIR;
+    public String ROOT_FOLDER_NAME = "RootFolder";
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: private please.
And [LCC](https://en.wikipedia.org/wiki/CamelCase#Variations_and_synonyms) for variables

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 15]
```diff
@@ -72,6 +78,51 @@ public void changed(ObservableValue observable, Object oldValue,
...
+            }
+            initTreeTableView();
+        });
+    }
+
+    private void initTreeTableView() {
+
+        Account account = clientConfiguration.getAccount();
+        AccountingHTTP http = null;
+        KEY_PAIR = clientConfiguration.getSelectedIdentity().getPrimaryKeyPair();
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: LCC

File: src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java [hunk 16]
```diff
@@ -72,6 +78,51 @@ public void changed(ObservableValue observable, Object oldValue,
...
+        try {
+            http = new AccountingHTTP(new AccountingServer(new URL(account.getProvider()).toURI(), account.getUser(), account.getAuth()), new AccountingProfile());
+        } catch (URISyntaxException | MalformedURLException e) {
+            e.printStackTrace();
+        }
+
+        try {
+            http.login();
+            http.updatePrefixes();
+            if(http.getProfile().getPrefixes().size() == 0){
```
Comment (julianseeger) [inline src/main/java/de/qabel/desktop/ui/remotefs/RemoteFSController.java]: hint: isEmpty()

File: src/main/resources/de/qabel/desktop/ui/remotefs/remotefs.fxml [hunk 17]
```diff
@@ -18,6 +18,11 @@
 
     <children>
       <HBox prefHeight="100.0" prefWidth="200.0">
+          <Button mnemonicParsing="false" onAction="#handleDownloadButtonAction" text="Download">
```
Comment (julianseeger) [inline src/main/resources/de/qabel/desktop/ui/remotefs/remotefs.fxml]: hint: if you add css `-fs-spacing: 10px` all buttons in the HBox get their insets automatically

File: src/main/resources/de/qabel/desktop/ui/remotefs/remotefs.fxml [hunk 18]
```diff
@@ -18,6 +18,11 @@
...
+              <HBox.margin>
+                  <Insets bottom="10.0" left="20.0" right="10.0" top="10.0" />
+              </HBox.margin>
+          </Button>
```
Comment (julianseeger) [inline src/main/resources/de/qabel/desktop/ui/remotefs/remotefs.fxml]: invalid fxml: the `<Button>` must be inside the `<childen>`

File: src/test/java/de/qabel/desktop/ui/remotefs/RemoteFSControllerTest.java [hunk 19]
```diff
@@ -66,9 +70,11 @@ public void setUp() throws Exception {
     @After
     public void after() throws Exception {
         localWrite.delete(localStorageFile.getName());
-        File dir = new File("tmp/testFolder");
-        FileUtils.deleteDirectory(dir);
-
+        FileUtils.deleteDirectory(new File(TMP_DIR + "/" + TEST_FOLDER));
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/remotefs/RemoteFSControllerTest.java]: hint:
```JAVA
new File(TMP_DIR + "/" + TEST_FOLDER))
```
is the same as
```JAVA
File tmpDir = new File(TMP_DIR);
new File(tmpDir, TEST_FOLDER);
```
especially useful if tmpDir is a class property (reused). This is semantically preferable over string concatenations.

File: src/test/java/de/qabel/desktop/ui/remotefs/RemoteFSControllerTest.java [hunk 20]
```diff
@@ -197,7 +200,7 @@ public void testUploadedDirectoryInChildNode() throws QblStorageException, IOExc
     @Test
     public void testCreateFolder() throws QblStorageException {
 
-        File dir = new File("tmp/testFolder");
+        File dir = new File(TMP_DIR +"/" + TEST_FOLDER);
```
Comment (julianseeger) [inline src/test/java/de/qabel/desktop/ui/remotefs/RemoteFSControllerTest.java]: pretty redundant. If all tests use the same dir, define it in a setUp method or extract a field

**Pull request merged**